document/core/exec/instructions.rst

```diff
@@ -2683,123 +2683,184 @@ Control Instructions
 
-7. For each catch clause :math:`(\CATCH~x_i~\instr_{2i}^\ast)` do:
+.. math::
+   ~\\[-1ex]
+   \begin{array}{r}
+   F; \val^m~(\TRYTABLE~\X{bt}~\catch^\ast~\instr^\ast~\END
+   \quad \stepto \quad
+   F; \HANDLER_n\{\catch^\ast\}~(\LABEL_n\{\epsilon\}~\val^m~\instr^\ast~\END)~\END \\ \qquad\qquad
+   (\iff \expand_F(\X{bt}) = [t_1^m] \to [t_2^n] \land (F.\AMODULE.\MITAGS[x]=a_x)^\ast)
+   \end{array}
+
+
+.. _exec-throw:
+
+:math:`\THROW~x`
+................
+
+1. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+
+2. Assert: due to :ref:`validation <valid-throw>`, :math:`F.\AMODULE.\MITAGS[x]` exists.
+
+3. Let :math:`a` be the :ref:`tag address <syntax-tagaddr>` :math:`F.\AMODULE.\MITAGS[x]`.
 
-   a. Assert: due to :ref:`validation <valid-try-catch>`, :math:`F.\AMODULE.\MITAGS[x_i]` exists.
+4. Assert: due to :ref:`validation <valid-throw>`, :math:`S.\STAGS[a]` exists.
 
-   b. Let :math:`a_i` be the tag address :math:`F.\AMODULE.\MITAGS[x_i]`.
+5. Let :math:`\X{ti}` be the :ref:`tag instance <syntax-taginst>` :math:`S.\STAGS[a]`.
 
-   c. Let :math:`H_i` be the handler :math:`(a_i~\instr_{2i}^\ast)`.
+6. Let :math:`[t^n] \toF [{t'}^\ast]` be the :ref:`tag type <syntax-tagtype>` :math:`\X{ti}.\TAGITYPE`.
 
-8. If there is a catch all clause :math:`(\CATCHALL~\instr_3^\ast)`, then:
+7. Assert: due to :ref:`validation <valid-throw>`, there are at least :math:`n` values on the top of the stack.
 
-    a. Let :math:`H'^?` be the handler :math:`(\epsilon~\instr_3^\ast)`.
+8. Pop the :math:`n` values :math:`\val^n` from the stack.
 
-9. Else:
+9. Let :math:`\X{exn}` be the :ref:`exception instance <syntax-exninst>` :math:`\{ \EITAG~a, \EIFIELDS~\val^n \}`.
 
-    a. Let :math:`H'^?` be the empty handler :math:`\epsilon`.
+10. Let :math:`\X{ea}` be the length of :math:`S.\SEXNS`.
 
-10. Let :math:`H^\ast` be the concatenation of :math:`H_i` and :math:`H'^?`.
+11. Append :math:`\X{exn}` to :math:`S.\SEXNS`.
 
-11. :ref:`Enter <exec-handler-enter>` the block :math:`\val^m~\instr_1^\ast` with label :math:`L` and exception handler :math:`H^\ast`.
+12. Push the value :math:`\REFEXNADDR~\X{ea}` to the stack.
+
+13. Execute the instruction |THROWREF|.
 
 .. math::
    ~\\[-1ex]
-   \begin{array}{l}
-   F; \val^m~(\TRY~\X{bt}~\instr_1^\ast~(\CATCH~x~\instr_2^\ast)^\ast~(\CATCHALL~\instr_3^\ast)^?~\END
-   \quad \stepto \\
-   \qquad F; \LABEL_n\{\epsilon\}~(\HANDLERadm_n\{(a_x~\instr_2^\ast)^\ast~(\epsilon~\instr_3^\ast)^?\}~\val^m~\instr_1^\ast~\END)~\END \\
-   (\iff \expand_F(\X{bt}) = [t_1^m] \to [t_2^n] \land (F.\AMODULE.\MITAGS[x]=a_x)^\ast)
+   \begin{array}{lclr@{\qquad}l}
+   S; F; \val^n~(\THROW~x) &\stepto& S'; F; (\REFEXNADDR~|S.\SEXNS|)~\THROWREF &
+     (\begin{array}[t]{@{}r@{~}l@{}}
+      \iff & F.\AMODULE.\MITAGS[x] = a \\
+      \land & S.\STAGS[a].\TAGITYPE = [t^n] \toF [] \\
+      \land & \X{exn} = \{ \EITAG~a, \EIFIELDS~\val^n \} \\
+      \land & S' = S \with \SEXNS = S.\SEXNS~\X{exn} ) \\
```

Comment:
Is it necessary to allocate exceptions in the store? Since they're immutable, can the abstract semantics not just treat them as ordinary values?

Reply:
References need internal identity, because it can become observable in the host, e.g. in JS. This is analogous to functions, which are likewise immutable.

(The lack of exn identity was one of the problems we faced with the JS API spec in the old proposal.)